### PR TITLE
Impl arithmetics for `NonZero<SecretScalar>`

### PR DESCRIPTION
**Pull request by survived**

### CODE REVIEW

File: generic-ec/src/arithmetic.rs [hunk 1]
```diff
@@ -345,22 +347,38 @@ impl_binary_ops! {
     Add (SecretScalar<E>, add, NonZero<Scalar<E>> = Scalar<E>) scalar::add,
     Add (NonZero<Scalar<E>>, add, SecretScalar<E> = Scalar<E>) scalar::add,
 
+    Add (NonZero<SecretScalar<E>>, add, Scalar<E> = Scalar<E>) scalar::add,
+    Add (Scalar<E>, add, NonZero<SecretScalar<E>> = Scalar<E>) scalar::add,
+    Add (NonZero<SecretScalar<E>>, add, NonZero<Scalar<E>> = Scalar<E>) scalar::add,
+    Add (NonZero<Scalar<E>>, add, NonZero<SecretScalar<E>> = Scalar<E>) scalar::add,
+
```
Comment (maurges) [inline generic-ec/src/arithmetic.rs:399]: Can update the comment above to mention NonZero<SecretScalar> as well

File: generic-ec/src/arithmetic.rs [hunk 2]
```diff
@@ -10,6 +10,9 @@ mod laws {
     };
     use crate::{Point, Scalar};
 
+    pub trait AlwaysNonZero {}
+    impl<T> AlwaysNonZero for NonZero<T> {}
```
Comment (maurges) [inline generic-ec/src/arithmetic.rs:14]: Missing docs

Also, why not make this a trait with methods? Maybe add something like `invert` and `safe_divide`

Comment (maurges) [reply generic-ec/src/arithmetic.rs:14]: Ah-huh, I completely missed the "laws" context. Never seen this module

Comment (survived) [reply generic-ec/src/arithmetic.rs:14]: This is internal trait, it's not exposed in API

Comment (survived) [reply generic-ec/src/arithmetic.rs:14]: trait is public, but the module is private

File: generic-ec/src/non_zero/mod.rs [hunk 3]
```diff
@@ -98,6 +98,67 @@ impl<E: Curve> NonZero<Scalar<E>> {
...
+        // it's actually non-zero, `CtOption` never exposes it, so `NonZero` with
+        // zero value is not accessible by anyone
+        CtOption::new(Self::new_unchecked(secret_scalar), is_non_zero)
+    }
+
+    /// Returns scalar inverse $S^{-1}$
+    ///
+    /// Similar to [Scalar::invert], but this function is always defined as inverse is defined for all
+    /// non-zero scalars
+    pub fn invert(&self) -> NonZero<SecretScalar<E>> {
```
Comment (maurges) [inline generic-ec/src/non_zero/mod.rs]: Should refer to SecretScalar::invert I guess

File: generic-ec/src/arithmetic.rs [hunk 4]
```diff
@@ -337,30 +339,46 @@ impl_binary_ops! {
     Mul (Scalar<E>, mul, Generator<E> = Point<E>) laws::mul_of_scalar_at_generator_is_valid_point,
 }
 
-// Point <> SecretScalar, Scalar <> SecretScalar, NonZero<Point> <> SecretScalar,
-// NonZero<Scalar> <> SecretScalar arithmetic ops
+// SecretScalar and NonZero<SecretScalar> multiplication, addition, substruction with Scalar,
+// NonZero<Scalar>, Point, and NonZero<Point>
 impl_binary_ops! {
```
Comment (maurges) [inline generic-ec/src/arithmetic.rs:389]: Damn these are hard to follow; and also they are a combinatorial explosion. In the future I want to format them as a table somehow

**Review by maurges — DISMISSED** [review]

File: generic-ec/src/arithmetic.rs (hunk 4 above)
Comment (survived) [reply generic-ec/src/arithmetic.rs:389]: Yeah, it's hard to read for me too. I'm usually looking at this expression that helps me to make sure that everything is implemented:

https://github.com/dfns/generic-ec/blob/c0ec6d016f95db5844fd4df96399c73b98ea15a0/generic-ec/src/arithmetic.rs#L557-L612

**Comment by survived** [comment]
@maurges I found a few arith trait not implemented, added them too in c0ec6d016f95db5844fd4df96399c73b98ea15a0 and 583aa5f28a871f93d0a648a2af5277de14fe07cb

**Comment by survived** [comment]
583aa5f28a871f93d0a648a2af5277de14fe07cb shows how easy it's to make an error! I was lucky clippy noticed that.

File: generic-ec/src/arithmetic.rs [hunk 5]
```diff
@@ -432,13 +477,42 @@ impl_unary_ops! {
...
+
     Scalar<E>, MulAssign, Scalar<E>, mul_assign, *,
+    Scalar<E>, MulAssign, NonZero<Scalar<E>>, mul_assign, *,
+    Scalar<E>, MulAssign, SecretScalar<E>, mul_assign, *,
+    Scalar<E>, MulAssign, NonZero<SecretScalar<E>>, mul_assign, *,
+
+    NonZero<Point<E>>, MulAssign, NonZero<Scalar<E>>, mul_assign, *,
+    NonZero<Point<E>>, MulAssign, NonZero<SecretScalar<E>>, mul_assign, *,
+    NonZero<Scalar<E>>, MulAssign, NonZero<Scalar<E>>, mul_assign, *,
+    NonZero<Scalar<E>>, MulAssign, NonZero<SecretScalar<E>>, mul_assign, *,
```
Comment (maurges) [inline generic-ec/src/arithmetic.rs:515]: Missing NonZero<Point> by generator.

Also missing a bunch of other NonZeros in left hand positions

**Review by maurges — APPROVED** [review]

**Comment by maurges** [comment]
Also for the future, to reduce a chance of errors, we could add tests that fuzz some algebraic properties

**Comment by survived** [comment]
There are some tests [here](https://github.com/dfns/generic-ec/blob/m/tests/tests/curves.rs), but they aren't comprehensive. We should do more.
